andela/ah-robotics
# #162949087 user can filter articles

### What does this PR do?
Allows a user to filter articles

### Description of Task to be completed?

- Allows a user to filter articles based on author, title and tags
- [x] Search and filter articles by `author`
- [x] Search and filter articles by `title`
- [x] Search and filter articles by `tags`

### How should this be manually tested?
1. Clone this repository on your local machine using 

     `git clone https://github.com/andela/ah-robotics.git`
2. Change your working directory to the project's directory using 
     
     `cd ah-robotics`
3. Change to the specific branch `ft-article-search-filter-162949087` using 

     `git checkout ft-article-search-filter-162949087`
4. Create a virtual environment using the following command: 
     
     `python3 -m virtualenv env`
5. Activate the virtual environment using 
     
     `source env/bin/activate`
6. Install all the requirements using 
     
     `pip install -r requirements.txt`

    #### Database setup

    Ensure [PostgreSQL](https://www.postgresql.org/) is installed and is running, follow these steps to create a new user and a database which will be used in the application.

    1. Start the postgres database server using the command:
      
        `pg_ctl -D /usr/local/var/postgres start`
    
    2. Run psql interactive terminal
    
        `psql postgres`
    
    3. Create a new user:
    
        `CREATE USER sample_username WITH PASSWORD 'sample_password';`
    
    4. Grant privileges to the user:
    
        `ALTER USER sample_username CREATEDB;`
    
    5. Create a database:
    
        `CREATE DATABASE sample_database_name WITH OWNER sample_username;`
    
    6. Exit psql interactive with the following command: `\q`
    
7. Create a **.env** file and add the following configuration variables

    ```source-json
          source env/bin/activate
          export DB_NAME="database_name"
          export DB_USER="sample_username"
          export DB_PASS="sample_password"
          export DB_HOST="host_name_eg_localhost"
          export DB_PORT="port_number_eg_5432"
          export TIME_DELTA="30"
          export EMAIL_HOST="smtp.gmail.com"
          export EMAIL_PORT=587
          export EMAIL_HOST_USER="user@example.com"
          export EMAIL_HOST_PASSWORD="email_password"
          export PASSWORD_RESET_URL="api/v1/reset_password"
          export CLIENT_DOMAIN="127.0.0.1:8000"
          export CLOUDINARY_NAME="your_cloudinary_name"
          export CLOUDINARY_KEY="your_cloudinary_key"
          export CLOUDINARY_SECRET="your_secret_key"
    ```
    
    > **Note**: The values of some of these variables are private namely the EMAIL_HOST_USER, EMAIL_HOST_PASSWORD and the CLOUDINARY VALUES.  Use your GMAIL email address and password for the  EMAIL_HOST_USER and EMAIL_HOST_PASSWORD respectively. To get Cloudinary values, create an account with Cloudinary and obtain the values.

8. Source .env to set the configuration variables in the terminal:

      `source .env`
9. Run migrations through the terminal to create the tables in the database:
      
      (i) Make the migrations using: 
      
      `python manage.py makemigrations`
      
      (ii) Apply the migrations using: 
      
      `python manage.py migrate`

10. Run the following command to collect static files.

    `python manage.py collectstatic`
      
#### Testing the features

1. Run the server from the terminal using: 
    
    `python manage.py runserver`
2. Download [Postman](https://www.getpostman.com/) and after it is installed, open it on your local machine and Paste the URL `http://127.0.0.1:8000/api/v1/users/` to first sign up a user.
3. On the **Headers** section add **Content-type** as the KEY and **_application/json_** as the VALUE.
4. Head over to the **Body** section and select the **raw** option and the add the following payload to the body:

    ```
        {
            "user": {
                "username": "sample_username",
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```

5. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
6. Log in the user using the following endpoint `http://127.0.0.1:8000/api/v1/users/login/` and insert the following payload to the body section.
    
    ```
        {
            "user": {
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```
7. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
8. After logging in the screen should look as follows:
    
    
<img width="1122" alt="screenshot 2019-01-29 at 16 26 20" src="https://user-images.githubusercontent.com/6994982/51911359-a76c8980-23e2-11e9-9e0c-53b32a9862ce.png">


9. Copy the token and click on the **Headers** section. In the **Key** column, type in the word **Authorization**. And in its corresponding value in the **Value** column, write the word __token__, add a space and paste in the token that you copied from the login response. The **Headers** section should look as follows after doing so:
    
    ![image](https://user-images.githubusercontent.com/10106044/51801938-67c46700-2255-11e9-91f6-31f2b5a5af46.png)

10. To Create an article, insert the following URL: http://127.0.0.1:8000/api/v1/articles/. In the body section, ensure the type is _raw_ and the format is _JSON(application/json)_. Finally, post an article by using the following payload:
    
    ```
    {
        "article": {
            "title": "Sample Article",
            "description": "Sample Article Description",
            "body": "The sample article body",
            "tagList": ["sample", "article"]
        }
    }
    ```
    Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.



11. To search and filter specific articles: 
- change the request to **GET**
- use the search articles endpoint and add the search and filter parameters to the endpoint as follows:  <http://127.0.0.1:8000/api/v1/article/search/?tagList=sample&author=sample_username&title=Sample>

This should return a list of filtered articles based on the passed parameters

<img width="1120" alt="screenshot 2019-01-30 at 20 55 16" src="https://user-images.githubusercontent.com/6994982/52001812-62735080-24d1-11e9-9730-1e57060f4f58.png">



#### Running tests
1. Head over to the terminal and ensure that the server is not running and you are in the project root folder.
2. Run the tests by typing in the following command: 
    
    `python manage.py test`

### Any background context you want to provide?
None

### What are the relevant pivotal tracker stories?
[#162949087](https://www.pivotaltracker.com/n/projects/2235171/stories/162949087)

### Screenshots
1. searching for an article

<img width="1120" alt="screenshot 2019-01-30 at 20 55 16" src="https://user-images.githubusercontent.com/6994982/52001870-86369680-24d1-11e9-8da2-06729fcea79a.png">



## Questions:
Questions and comments are highly welcome.
